### PR TITLE
[@types/react] add VoidFunctionComponent type which does not accept "children"

### PR DESCRIPTION
**Pull request by awmottaz**
Adds a type `VoidFunctionComponent` with `VFC` alias that is the same as
the `FunctionComponent` (`FC`) type, except that it does not accept a
`children` prop.

cf. #40993

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/40993#issuecomment-607370492
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

File: types/react/test/index.ts
Comment (eps1lon) [inline types/react/test/index.ts:270]: Since this component isn't supposed to accept children wouldn't the error be desired?

**Review by eps1lon — CHANGES_REQUESTED** [review]
Big fan of the idea. What do you consider void about this component?

I think most the types maintainers agree that this should be the default in new major release of `@types/react` so we might want to reflect that in the name.

File: types/react/test/index.ts
Comment (awmottaz) [reply types/react/test/index.ts:270]: Ah yes, sorry about the confusing verbiage here

**Comment by awmottaz** [comment]
> What do you consider void about this component?

The naming is inspired by the analogy of "normal element" → `FunctionComponent`, so "void element" → `VoidFunctionComponent`
https://html.spec.whatwg.org/multipage/syntax.html#elements-2

**Comment by awmottaz** [comment]
> I think most the types maintainers agree that this should be the default in new major release of @types/react so we might want to reflect that in the name.

Do you mean that the `FunctionComponent` type itself should be changed to what I have called `VoidFunctionComponent`?

I could use some help with the tests. It's not clear to me what is causing the failure, and I saw failures when I ran the tests before adding my changes.

**Comment by eps1lon** [comment]
> Do you mean that the FunctionComponent type itself should be changed to what I have called VoidFunctionComponent?

Yes, we generally advise for that specific reason against using `React.FunctionComponent`. But removing implicit `children` would likely break a lot of things so we want to wait for React 17.

**Comment by awmottaz** [comment]
> Yes, we generally advise for that specific reason against using `React.FunctionComponent`. 

Just curious where that advice has been given? Maybe you can comment on #40993?

> But removing implicit children would likely break a lot of things so we want to wait for React 17.

Yes, there is quite a lot that would probably break. For that reason, I think that a new type might be pragmatic and get something more useful to developers sooner and avoid the refactoring lift later. That change would force every `FunctionComponent` to define its own `children` prop and could be very painful for some projects. But I'm happy to defer to the maintainers of this package and would love to hear the counter-arguments in favor of waiting for v17 and modifying `FunctionComponent` instead of introducing a new type.

**Comment by eps1lon** [comment]
> Just curious where that advice has been given?

Informal in https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/function_components#docsNav, various conversations in this repo etc. Though we probably should add a warning to the JSDOC and recommend `VoidFunctionComponent` + explicit children.

File: types/react/index.d.ts
Comment (eps1lon) [inline types/react/index.d.ts:82]: This might cause CI to fail? Or is `master` currently broken?

**Review by eps1lon — CHANGES_REQUESTED** [review]
Agree with the change and the naming makes sense to me now. React also uses that terminology:

> input is a void element tag and must neither have `children` nor use `dangerouslySetInnerHTML`.
 
We need to figure out what's wrong with CI though.

File: types/react/index.d.ts
Comment (awmottaz) [reply types/react/index.d.ts:82]: `master` seems to be currently broken

**Comment by eps1lon** [comment]
@awmottaz Rebasing onto `master` should this PR go green now.

**Comment by awmottaz** [comment]
Confirmed that with latest `master` the tests pass. I'm looking into these new test failures.

File: types/react/index.d.ts
Comment (awmottaz) [reply types/react/index.d.ts:82]: Thanks, this does seem to be the problem. I'm trying it out now with just using VoidFunctionComponent here. Does this look okay to you?

```suggestion
    type ComponentType<P = {}> = ComponentClass<P> | VoidFunctionComponent<P>;
```

Comment (awmottaz) [reply types/react/index.d.ts:82]: Ah, this is no good either. Going to try by reverting this line without `VoidFunctionComponent`.

**Comment by ferdaber** [comment]
@eps1lon would we be looking to deprecate VFC on the next major? Or just pointing the FC type to VFC?

**Comment by eps1lon** [comment]
> would we be looking to deprecate VFC on the next major? Or just aliasing the FC type to VFC?

Replace FC with the definition of VFC, alias VFC to FC and deprecate VFC. Basically FC becomes VFC at which point VFC is no longer needed so we deprecate it.

**Comment by awmottaz** [comment]
@eps1lon I simplified this PR a lot. I was having trouble getting tests to pass whenever I used VFC in other types that also reference FC. I'm guessing that it's because TypeScript cannot always infer whether the component is a FC vs a VFC.

Hopefully this stripped down change still provides some value!

**Review by eps1lon — APPROVED** [review]

File: types/react/test/tsx.tsx
Comment (eps1lon) [inline types/react/test/tsx.tsx:34]: This looks like it doesn't belong here.

**Review by eps1lon — CHANGES_REQUESTED** [review]

**Review by eps1lon — APPROVED** [review]

**Review by ferdaber — APPROVED** [review]

**Comment by awmottaz** [comment]
Ready to merge

**Comment by ljharb** [comment]
This is a really confusing choice of name. "void x" usually means "an x that returns nothing", and "accepting children" has no correlation to whether the component renders something or not.

Is it too late to perhaps provide a better name (leaving the old one as an alias as needed, for compat)? Perhaps, like HTML, an element that takes no children should be called "self-closing"?

**Comment by jahglow** [comment]
Yeah, I wonder how this PR was merged with such a name, a disgrace to React Community *commits a harakiri*

**Comment by ljharb** [comment]
cc @eps1lon re https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46643#issuecomment-771679165

**Comment by eps1lon** [comment]
We settled on "void component" since that is established terminology in React (e.g. `<input>foo</input>` warns about `input is a void element tag`) which is used because it is established terminology in HTML itself:

> Void elements can't have any contents (since there's no end tag, no content can be put between the start tag and the end tag).

-- https://html.spec.whatwg.org/multipage/syntax.html#elements-2

So it does make sense to me to call them "VoidFunctionComponent" in the react domain. But I can see that this might clash with how we use "void" in javascript itself?

We don't intend to keep the name around anyway. `VoidFunctionComponent` will become `FunctionComponent` (probably once React 18 lands) i.e. typings will no longer declare implicit children. 

Maybe I'm misinterpreting the HTML spec but if you agree that "void" does make sense with respect to HTML I'd rather stick with the name (to not create even more type names) since it's temporary anyway.

**Comment by ljharb** [comment]
@eps1lon right but that's because it doesn't *render* its children. It still accepts them. This type is about _props_ not rendering.

**Comment by eps1lon** [comment]
> Components can accept or not accept children, and that has zero guaranteed correlation to whether they render children or not.

If a component does not accept children it cannot render them. That is a guarantee. But I understand that components that do accept children do not guarantee that they render them e.g. `<select />`.

Could you clarify how a `<input>` void element is different from a `Input` void function component? Neither accept children.

**Comment by ljharb** [comment]
Incorrect:
```jsx
function RendersChildren() {
  return <div>children</div>;
}
```

**Comment by eps1lon** [comment]
> Incorrect:

Which part is incorrect? 

`RendersChildrenButDoesNotAcceptChildren` does not accept children and therefore doesn't render them. It does however render something. I didn't mean to imply that `VoidFunctionComponent`s render nothing. Just like "void elements" don't render nothing.

**Comment by ljharb** [comment]
It renders children because the element it produces has children. "renders children" does not mean "renders the children prop it receives".

```jsx
function RendersChildrenAndAcceptsChildren({ children }) {
  return <div>different children</div>
}
```

Void elements render _one thing_ with no children. For custom components, that is entirely decoupled from whether it accepts children as props or not.

**Comment by esebastian** [comment]
I find it hard to follow your point of view @ljharb, because for me this is not about what the component is rendering (that's precisely what the component is abstracting), but about how it's being used: ```<MyComponent />``` vs ```<MyComponent>Some children</MyComponent>```.

**Comment by ljharb** [comment]
@esebastian i totally agree, that's why i think "void element" is the wrong term - because that describes *rendering behavior* not interface behavior, and the type being used here describes the interface behavior. `<MyComponent />` passes no children, `<MyComponent>Some children</MyComponent>` and `<MyComponent children="some children" />` both do, and the type referenced in this PR has absolutely nothing to do with the rendering behavior of MyComponent - thus, it should have a distinct term from one that describes the rendering behavior.

**Comment by mvasin** [comment]
Agree that the word `void` is very confusing and should be about the return data type, not the input data type. But seems like this naming confusion set in from the [HTML spec](https://html.spec.whatwg.org/multipage/syntax.html#void-elements):

> Void elements only have a start tag; end tags must not be specified for void elements.

The quote above is about the input data type - I can't _pass_ anything between the opening and closing tags to the HTML element.

**Comment by flying-sheep** [comment]
cross-posting https://github.com/yannickcr/eslint-plugin-react/issues/2913#issuecomment-996646581:

> I like that they’ll delete `FC` and rename `VFC`→`FC` in the next major release. That’s how `FC` should have been defined from the start.
> 
> If that’s the intention, `VFC` is truly a horrible name though. it should be named `E`(xplicit)`FC` then: “VFC” implies the upgrade path is:
> 
>     * use FC for components taking children, specify child type in props interface only if it’s not `ReactNode`
> 
>     * use VFC for components not taking children
> 
>     * sort out the mess when upgrading to the next major typing version
> 
> 
> I much prefer this upgrade path:
> 
>     * use `import { VFC as FC } from 'react'` everywhere, explicitly state child types (or lack thereof) everywhere
> 
>     * (optional) add this to your `.eslintrc.yaml`:
>       ```yaml
>       no-restricted-imports:
>       - error
>       - paths:
>         - name: react
>           importNames: [FC]
>           message: 'Use VFC instead and directly type children (e.g. `children?: ReactNode`)'
>       ```
> 
>     * delete `VFC as ` (and eslint rule) in next major typing version

**Pull request merged**
